app/uk/gov/hmrc/gform/playcomponents/GformSessionCookieCryptoFilter.scala

```diff
@@ -0,0 +1,90 @@
+/*
+ * Copyright 2021 HM Revenue & Customs
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package uk.gov.hmrc.gform.playcomponents
+
+import _root_.akka.stream.Materializer
+import cats.syntax.eq._
+import com.typesafe.config.Config
+import play.api.mvc._
+import play.api.routing.Router.RequestImplicits._
+import uk.gov.hmrc.crypto.{ Crypted, Decrypter, Encrypter, PlainText }
+import uk.gov.hmrc.gform.FormTemplateKey
+import uk.gov.hmrc.gform.controllers.CookieNames._
+import uk.gov.hmrc.gform.gformbackend.GformConnector
+import uk.gov.hmrc.gform.sharedmodel.formtemplate.{ Anonymous, FormTemplate, FormTemplateId }
+import uk.gov.hmrc.http.HeaderCarrier
+import uk.gov.hmrc.play.HeaderCarrierConverter
+import uk.gov.hmrc.play.bootstrap.frontend.filters.crypto.{ SessionCookieCrypto, SessionCookieCryptoFilter }
+
+import scala.concurrent.{ ExecutionContext, Future }
+
+class GformSessionCookieCryptoFilter(
+  sessionCookieCrypto: SessionCookieCrypto,
+  nonanonymouscookieCryptoFilter: SessionCookieCryptoFilter,
+  anonymousCookieCryptoFilter: SessionCookieCryptoFilter,
+  val sessionBaker: SessionCookieBaker,
+  gformConnector: GformConnector,
+  config: Config
+)(implicit override val ec: ExecutionContext, override val mat: Materializer)
+    extends SessionCookieCryptoFilter {
+
+  override protected lazy val encrypter: Encrypter = sessionCookieCrypto.crypto
+  override protected lazy val decrypter: Decrypter = sessionCookieCrypto.crypto
+
+  private val AnonymousAuthConfig = "anonymous"
+  private val NonAnonymousAuthConfig = "nonanonymous"
+  private val FormTemplateIdClass = "uk.gov.hmrc.gform.sharedmodel.formtemplate.FormTemplateId"
+
+  override def apply(next: RequestHeader => Future[Result])(rh: RequestHeader): Future[Result] = {
+
+    lazy val authconfigCookieValue = Crypted(rh.cookies.get(authConfigCookieName).fold("")(_.value))
+
+    val isPathParameterFormTemplateIdExists = rh.handlerDef.exists { hd =>
+      hd.parameterTypes.exists { _.getName() === FormTemplateIdClass }
+    }
+
+    val maybeFormTemplate: Future[Either[Unit, FormTemplate]] =
+      if (isPathParameterFormTemplateIdExists) {
+        val templateId = rh.uri.split("/")(3)
```

Comment:
`FormTemplateId` is not guaranteed to be on index `3`. We need to calculate index from `handlerDef` data.